src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+        // rc2 is the last ref to the `__RefCount` created in this block -> gets freed
+    }
+    assert(rc.isUnique);
+}
+
+/**
+A qualified reference counted `struct` that is intended to be composed by
+user-defined types that desire to implement manual memory management by means of
+reference counting.
+
+`__RefCount` was designed to be composed as a field inside a user-defined type.
+The user is responsible to initialize the `__RefCount` in that type's constructors.
+The user will call the `isUnique()` method to decide if this is the last
+reference to the enclosing type so memory can be safely deallocated.
+
+Implementation: The internal implementation of `__RefCount` uses `malloc`/`free`.
+
+Important: The `__RefCount` member must be initialized through a call to its
+constructor before being used.
+*/
+struct __RefCount
+{
+    import core.atomic : atomicOp;
+
+    alias CounterType = uint;
+    private shared CounterType* rc = null;
+
+    /*
+    Perform `rc op val` operation. Always use atomics as the counter is shared.
+
+    Returns:
+         The result of the operation.
+    */
+    @nogc nothrow pure @trusted scope
+    private CounterType rcOp(this Q, string op)(CounterType val) const
+    {
+        return cast(CounterType)(atomicOp!op(*(cast(shared CounterType*) rc), val));
+    }
+
+    /**
+    Creates a new `__RefCount` instance. Its memory is internally managed with
+    malloc/free.
+
+    Params:
+         _ = an unused int value; required because structs don't have a
+             user defined default constructor
+    */
+    @nogc nothrow pure @trusted scope
+    this(this Q)(int _)
+    {
+        // We are required to always use a shared support as the result of a `pure`
+        // function is implicitly convertible to `immutable`.
+
+        shared CounterType* support = cast(shared CounterType*) pureAllocate(CounterType.sizeof);
+        *support = 1; // Start with 1 to avoid calling an `addRef` from 0 to 1
+        rc = cast(typeof(rc)) support;
+    }
+
+    private enum copyCtorIncRef = q{
+        rc = rhs.rc;
+        assert(rc == rhs.rc);
+        if (rhs.isInitialized())
+        {
+            addRef();
+        }
+    };
+
+    /**
+    Copy constructs a mutable `__RefCount` from a mutable reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs)
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a `shared` mutable `__RefCount` from a `shared` mutable
+    reference, `rhs`. This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope ref shared typeof(this) rhs) shared
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const __RefCount) from a mutable reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs) const
```

Comment:
Could this be `inout` to avoid some duplications?

Reply:
Yes, made the `const` ones `inout`.